src/main/java/javax/persistence/spi/PersistenceProviderResolverHolder.java

```diff
@@ -170,32 +163,30 @@ private void loadResolverClasses(ClassLoader cl) {
 				//problem #1: avoid hard ref with classloader (List<WR<PP>>?
 				//problem #2: avoid half GC lists
 				// todo (steve) : why arent we just caching the PersistenceProvider *instances* as the CachingPersistenceProviderResolver state???
-				synchronized ( resolverClasses ) {
-					List<PersistenceProvider> providers = new ArrayList<PersistenceProvider>( resolverClasses.size() );
+				List<WeakReference<Class<? extends PersistenceProvider>>> updatedResolverClasses = cachedResolverClasses.get();
+				do {
+					List<PersistenceProvider> providers = new ArrayList<PersistenceProvider>(updatedResolverClasses.size());
 					try {
-						for ( WeakReference<Class<? extends PersistenceProvider>> providerClass : resolverClasses ) {
-							providers.add( providerClass.get().newInstance() );
+						for (WeakReference<Class<? extends PersistenceProvider>> providerClass : updatedResolverClasses) {
+							providers.add(providerClass.get().newInstance());
 						}
-					}
-					catch ( InstantiationException e ) {
```

Comment:
both throw exception and return statement should be handled outside the doWhile Loop